### DISCUSSION TITLE
How can I make code wrap when generating PDF from Jupyter Notebooks?

### DISCUSSION THREAD

**Discussion by entelecheia1963** [quarto-dev/quarto-cli#7612] [Q&A]
### Description

I am trying to generate PDF files from a Jupyter notebook within vscode. The code in the generated PDF file does not wrap - but overflows the bounding box even though I have the 'code-overflow: wrap' set.

Unfortunately I can't upload the generating .ipynb file here - so I have renamed the file with a .txt extension

[code-wrap-issue.pdf](https://github.com/quarto-dev/quarto-cli/files/13382472/code-wrap-issue.pdf)
[code wrap issue.ipynb.txt](https://github.com/quarto-dev/quarto-cli/files/13382479/code.wrap.issue.ipynb.txt)

I have only two cells in my workbook:

A 'raw' cell that is:

```
---
title : DIY Model Fitting
author : Dr Rob Collins
date: 16/11/2023
format: 
  pdf:
    toc: true
    number-sections: true
    colorlinks: true
  docx:
    toc: true
    number-sections: true
    highlight-style: github
highlight-style: zenburn
echo: true
code-overflow: wrap
---
```

And a python cell that is:

``` {python}
a = 2
b = 3

c = a * b  # But the problem is that if I have either a long line of code or long line of comments then wrapping does not seem to work

c = a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b 
```
Any ideas on how to fix this would be excellent. Thank you.

**Comment by mcanouil** [comment]
Thanks for the interest in Quarto.
Please do note that GitHub offers search abilities in discussions and in issues, where you could have seen several discussions on the matter such as:
- #4121
There are other discussions on the subject.

Note also that your code is not really usable, see <https://quarto.org/bug-reports.html>.
The following is:

````qmd
---
title: DIY Model Fitting
author: Dr Rob Collins
date: 16/11/2023
format: pdf
---

```{python}
a = 2
b = 3

c = a * b  # But the problem is that if I have either a long line of code or long line of comments then wrapping does not seem to work

c = a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b 
```
````

Also, you should read the documentation. `echo: true` should be under `execute` (<https://quarto.org/docs/computations/execution-options.html>).


<details><summary>One of the solution from the discussion linked.</summary>

````qmd
---
title: "Code wrap in LaTeX PDF"
format: pdf
include-in-header: 
  - text: |
      \usepackage{fvextra}
      \DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaklines,commandchars=\\\{\}}
include-before-body:
  - text: |
      \RecustomVerbatimEnvironment{verbatim}{Verbatim}{
        showspaces = false,
        showtabs = false,
        breaksymbolleft={},
        breaklines
        % Note: setting commandchars=\\\{\} here will cause an error 
      }
---

```{python}
a = 2
b = 3

c = a * b  # But the problem is that if I have either a long line of code or long line of comments then wrapping does not seem to work

c = a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b 
```
````

<img width="710" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/9a751340-3ea4-4671-a336-a975529371cb">

</details>

**Comment by entelecheia1963** [comment]
Thank you again for your kind support!

So I think you were saying that my code samples didn't have enought tick-marks or something? I'll certainly watch out for that in future.

I think that one of the many advantages you had over me was that knowing the slang expression 'folding' was being used in the discussio forums to mean 'wrap' (as is used in the keyword related to this issue). Again, thank you for educating me about that.

I think I see from those discussion you pointed out that code 'folding' does not, in fact, work for PDFs.

Thank you again!

**Reply by mcanouil** [reply]
I actually searched for `wrap label:latex` (https://github.com/quarto-dev/quarto-cli/discussions?discussions_q=wrap+label%3Alatex), *i.e.*, discussions labelled with `latex` and one keyword "wrap". There are only twelve discussions. I checked four of them and linked one.

The solution from the discussion I linked, is in my answer, you only have to "unfold" the details code.

<img width="349" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/1fc0c1f7-1a68-4915-af6c-80de60225869">


For the future, you can share a Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
---

This is a reproducible Quarto document using `format: html`.
It is written in Markdown and contains embedded R code.
When you run the code, it will produce a plot.

```{r}
plot(cars)
```

The end.
````
`````

**Comment by entelecheia1963** [comment]
I thought I had the answer with the above suggestion. However, whilst wrapping does seem to work .. unfortunately it now seems to corrupt other code blocks. It deletes some spaces .. but also in some other cells in my main project it seems to duplicate parts of the text.

I have been searching with the 'wrap label:latex' keywords - but have not found a solution as yet.

[code-wrap-issue.pdf](https://github.com/quarto-dev/quarto-cli/files/13399860/code-wrap-issue.pdf)

It looks like this (I was told off for including images previously - but this seems to be the clearest way of showing the issue): 
![image](https://github.com/quarto-dev/quarto-cli/assets/42785164/bf7b2e5b-a7f9-4b80-a98b-4725020e9f06)

My file looks like this (again, doing my best to match the style required above by copy-pasting from a Python notebook)

````qmd
---
title : Code not wrapping in PDF
author : Me
date: 16/11/2023
format: 
  pdf:
    toc: true
    number-sections: true
    colorlinks: true
    include-in-header: 
      text: |
        \usepackage{fvextra}
        \DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaklines,commandchars=\\\{\}}
        \DefineVerbatimEnvironment{OutputCode}{Verbatim}{breaklines,commandchars=\\\{\}}
highlight-style: zenburn
execute:
  echo: true
code-overflow: wrap
---

```{python}
a = 2
b = 3

c = a * b  # But the problem is that if I have either a long line of code or long line of comments then wrapping does not seem to work

c = a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b + a + b 

# I don't know how to show jupyter notebook cell boundaries in this notation - sorry

import pandas as pd
import numpy as np
```
````

**Reply by mcanouil** [reply]
The thing is that this has almost nothing to do with Quarto.

$\LaTeX$ is what it is ...

Here is the method to work/debut:
- use `keep-tex: true` the ne examine the $\LaTeX$ code.
- look on the web for $\LaTeX$ solution (stack overflow/exchange, etc.)
- use `include-in-header` and/or `include-before-body` Quarto option to include the solution you've found.

- and/or modify the template/partial https://quarto.org/docs/journals/templates.html#latex-partials

In the end, you are going to need to learn your way in $\LaTeX$ and learn it at least a bit if you want to customise the output.
